### ISSUE TITLE
Common problems with MVVM apps

### ISSUE THREAD

**Issue by pshomov** [pshomov/pshomov.github.io#3]

**Comment by pshomov** [comment]
What do you think guys, do you find these issues problematic or is it just me?

**Comment by georgejecook** [comment]
I don't think you have enough information here to start a discussion. You should expand on your points if you want a serious conversation about this.

**Comment by pshomov** [comment]
@georgejecook 

I do agree that I did not go deep enough into any of my points. The main reason for that is I did not want to write a big blog post, since I was hoping that once I point these issues out people would connect.

May be I will write another post that drills more into these.

Still, do you feel that these alleged problems are really non-issues for you?

**Comment by georgejecook** [comment]
I honestly don't think there's enough to make a conversation here. A blog post entitled "Common problems with MVVM apps", with 2 points, which are not explored in any depth, written by someone who confesses little knowledge of the pattern is not exactly a fire-starter.

It's fine to point out things that are a pain; but you shouldn't frame it as a insightful critique of MVVM. How do you honestly know what are common MVVM problems?

It might have been better to call it "teething problems with MVVM" and ask for advice, because as it's worded, you're likely to get "yes, me too!" posts from noobs, who won't help you, and those who are experienced will likely think similarly to myself which is "there is no discussion here, this guy's just starting out with MVVM".

I hope this comes accross as constructive, which is how it's intended. If you want advice: ask; If you want to critique and start a debate - be knowledgable and do it properly.

**Comment by pshomov** [comment]
@georgejecook May be I downplayed myself a bit too much, I think I am going to edit my post to clarify my experience. 

My main point was that having looked at and used UI patterns in Java/C#/JavaScript the MVVM apps seem lacking when it came to separation of concerns. Most often the viewmodels seem to contain everything between the view and the database connection and that I see as a problem. I will go so far to say I find this a violation of the description of the pattern - the view model is to be concerned with UI related logic only.

I have read and watched a lot connected with MVVM and have about 1 year of on/off writing MVVM apps(Xamarin Forms). I think that should be enough to get a feeling for what are the problems in MVVM apps. Sure, there might be other problems with MVVM but I see the issues above over and over. 

I guess the title could have been better :disappointed:

**Comment by georgejecook** [comment]
What you described is not MVVM. the vM should know nothing about the view, and any DAO/service layer should also be abstracted. It sounds like your gripe is with people who are using MVVM incorrectly. Can you point us to some code which you think is typical of the common problems you've identified?

**Comment by pshomov** [comment]
now we are talking ;) I will point again at the examples I had in my post, since I find them representative of the issues - [here](https://github.com/dotnetcurry/wpf-mvvmlight/blob/master/WPF_MVVMLight_CRUD/ViewModel/MainViewModel.cs), [here](https://github.com/MvvmCross/MvvmCross-Tutorials/blob/master/Sample%20-%20TwitterSearch/TwitterSearch.Core/ViewModels/TwitterViewModel.cs) and [here](https://github.com/rid00z/FreshMvvm/blob/master/samples/FreshMvvmSampleApp/FreshMvvmSampleApp/PageModels/ContactPageModel.cs)

The problems with these examples - 1)viewmodels should not be deciding **how** or **when** to interact with services 2) viewmodels should not be doing navigation (a concrete case of the first rule really).

BTW, I love this blogging thing ;) I learn so much trying to communicate what I mean.

**Comment by MichaelRumpler** [comment]
I think when you want to discuss about MVVM, then you should not mix up the terms yourself.
You wrote "Too much logic and responsibilities in the views", but you really meant "... in the view models".

I'm sure we all agree that the view should only be responsible for the UI. In MVVM, the view receives the data from the viewmodel and the viewmodel is the glue which connects everything together. So the viewmodel does access services and navigate to other viewmodels.
In some (bad) samples the viewmodels even access the database directly. The links you posted already do this better with services and DI.

If you don't like that, then you could also use MVC or MVP. The controllers and presenters have slightly different responsibilities which may suit you more.

A lengthy discussion about those patterns can be found in [this 2011 article about MVPVM](https://msdn.microsoft.com/en-us/magazine/hh580734.aspx) published in the MSDN magazine.

**Comment by pshomov** [comment]
Hi @MichaelRumpler 

You are right, I did not notice I mixed those up. I will fix this up.
I don't fully agree with this:

> In MVVM, the view receives the data from the viewmodel and the viewmodel is the glue which connects everything together

I am saying the viewmodel does not need to glue everything together. More importantly it does not have the _lifetime_ to glue everything together. I think we need to put the "glue together" kind of logic in separate object(App object) and let view models just take care of making the UI shine and send back to the App object the user's commands. The App object has the lifetime and level of overview to decide when and how things should be done. 

I admin I am not happy with my post and made an attempt to make myself better understood in [take two](http://pshomov.github.io/common-problems-with-mvvm-apps-take-two/). Hope I make my point there better.

**Comment by captainjono** [comment]
@pshomov 100% agree with your observations. As someone who has taken ownership of a project that was based on PRISM4(MVVM framework) i have seen first hand how unmanagable and coupled a solution can get when someone who doesnt understand MVVM properly gets a hold of it. And by not understanding, i mean someone who just follows example code without understanding the theory and central tenants of the pattern. I always find myself creating said AppObject to handle all the domain logic as required on the client, it not only to manage concurrency/network relates issues. If you cant take your viewmodel and and place various other views ontop of it, its the first sign you are not doing it proplerly. If you cant reuse your services to populate other viewmodels efficently, its another sign. 

I honestly dont see any other way to create maintainable complex applications at scale though with this pattern. Its all about seperating concerns, and these days im even comming around to the idea that two-way data binding is a bad idea because this leads people to think just because their DTO has been updated magically, they should be sending it back to the server via the viewmodel. Sure it may work for your CRUD exmaples, but if u were doing CRUD IMO MVVM is overkill. I am purist though. i keep my views expressed souly in markup with Commands/Actions/Behaviors, as opposed to managing state transitions with code-behind.

**Comment by pshomov** [comment]
@captainjono I'm glad you liked the post, might enjoy the follow up [here](http://pshomov.github.io/common-problems-with-mvvm-apps-take-two/) and I am planning on having a concrete example next week of my suggestion on how to fix this situation. Stay tuned ;-)

**Comment by pshomov** [comment]
For those who cared about this one way or another, this new post by me might be of interest where I propose how we deal with the issues at hand - http://pshomov.github.io/compartmentalizing-logic
